### PR TITLE
Skip hashing files if the mod time is before the last summary

### PR DESCRIPTION
**Pull request by angelini**
This PR includes a `LatestModificationTime` in all summary files.

This time is then used to skip computing hashes for certain files. If a file's modification time is less than the `LatestModificationTime` on a subsequent diff, we assume that the file has not changed and skip reading it and computing it's hash.

It is still possible that a file's mode has been changed - which will not update it's modification time - so we still need to handle that case.

There are three distinct cases and the performance impacts are different for each case:

- Running an initial diff on a folder to generate the first summary file. This case is unaffected by this PR as all files need to be read.
- Running a diff over a directory with no changes. This case is greatly sped up as all file hashing will be skipped. This case actually does happen in Gadget when we run blocks of FS code which may potentially write to disk - but often don't.
- Running a diff over a directory with some changes. In this case the speed up depends on how many files changed, it should only need to hash new files.

Data from local benchmarking over an example Gadget project: https://gist.github.com/angelini/9edf05e16ddb6180858c48d6c0adcd12

No changes case: ~80% improvement
3 new NPM dependencies case: ~80% improvement

### CODE REVIEW

**Review by arbourd — APPROVED** [review]
Code looks fine, but I don't know enough about fsdiff to catch anything useful here.

File: pkg/diff/diff.go [hunk 1]
```diff
@@ -138,23 +140,26 @@ func WalkChan(dir string, ignores []string) <-chan *Entry {
...
-				hash, err = hashLink(path)
-			} else {
-				hash, err = hashFile(path)
-			}
-			if errors.Is(err, fs.ErrNotExist) {
-				return nil
-			}
-			if err != nil {
-				return pushErr(fmt.Errorf("hash file: %w", err))
+			if info.ModTime().UnixNano() >= latestModTime {
```
Comment (airhorns) [inline pkg/diff/diff.go:143]: Do we need to run a fancy filesystem in production to get nanonsecond precision in the mtime?

File: pkg/diff/diff.go [hunk 2]
```diff
@@ -51,21 +51,23 @@ func hashEmptyDir() []byte {
...
 }
 
 func (e *Entry) toPb() *pb.Entry {
 	return &pb.Entry{
-		Path: e.path,
-		Mode: uint32(e.mode),
-		Hash: e.hash,
+		Path:    e.path,
+		Mode:    uint32(e.mode),
+		ModTime: e.modTime,
```
Comment (airhorns) [inline pkg/diff/diff.go:65]: We don't actually write out these pbs anywhere that is stored for a long time right? Like they are only ephemeral such that this change won't cause any forwards/backwards compatibility issues with already stored summaries? I guess since it is a field add it would also be ok if we did right?

File: test/diff_test.go [hunk 3]
```diff
@@ -381,7 +381,7 @@ func TestDiffWithEmptyDirectories(t *testing.T) {
 		"e/f": "f2",
 	}, []string{"b/c", "b/d"})
 
-	d2, s2, err := diff.Diff(diff.WalkChan(tmpDir, []string{}), diff.SummaryChan(s1))
+	d2, s2, err := diff.Diff(diff.WalkChan(tmpDir, []string{}, s1.LatestModTime), diff.SummaryChan(s1))
 	if err != nil {
 		t.Fatalf("failed to run diff: %v", err)
 	}
```
Comment (airhorns) [inline test/diff_test.go:388]: Can we contrive any tests that stress the timestamp comparison logic to capture some of the edge cases mentioned in the git design doc?

**Review by airhorns — COMMENTED** [review]
> No changes case: ~80% improvement
> 3 new NPM dependencies case: ~80% improvement

HUBBA HUBBA

File: pkg/diff/diff.go (hunk 1 above)
Comment (angelini) [reply pkg/diff/diff.go:143]: > Since kernel 2.5.48, the stat structure supports nanosecond resolution for the three file timestamp fields. 

This seems to depend more on the kernel than the filesystem implementation, since it's calling out to a kernel syscall.

And it will also depend on the language / runtime, for example in Python the base `st_mtime` is second precision and they needed to add `st_mtime_nano` for nanosecond support.

File: pkg/diff/diff.go (hunk 2 above)
Comment (angelini) [reply pkg/diff/diff.go:65]: Yeah they are only temporary, but after reviewing the PR we only need these temporarily in the `Entry` struct but could discard them before writing to disk.

Comment (angelini) [reply pkg/diff/diff.go:65]: That's right, all of the `fsdiff` files are super temporary in nature. If this was rolled out into production we could be a bit more careful about how we handle the new field, but there is nowhere that this will exist on disk for a long time.

File: test/diff_test.go (hunk 3 above)
Comment (angelini) [reply test/diff_test.go:388]: The tests - at least locally - already stress the `>=` clause by landing with the exact same nanosecond on some runs.

I'll add a couple new tests with different mod times passed here to properly capture how files are skipped with old mod times.
